framework/src/materials/MaterialPropertyStorage.C

```diff
@@ -361,30 +366,191 @@ MaterialPropertyStorage::initProps(MaterialData & material_data,
     {
       const auto prop_id = _stateful_prop_id_to_prop_id[i];
       mat_props.setPointer(i, material_data.props(0)[prop_id].clone(n_qpoints), {});
+      mooseAssert(mat_props[i].id() == prop_id, "Inconsistent id");
     }
 }
 
 void
 dataStore(std::ostream & stream, MaterialPropertyStorage & storage, void * context)
 {
-  unsigned int max_state = storage.maxState();
-  dataStore(stream, max_state, context);
+  auto num_states = storage.numStates();
+  dataStore(stream, num_states, nullptr);
+
+  // Store the material property ID -> name map for mapping back
+  const auto & registry = storage.getMaterialPropertyRegistry();
+  std::vector<std::string> ids_to_names(registry.idsToNamesBegin(), registry.idsToNamesEnd());
```

Comment:
Can't due to non-cost `dataStore(std::istream &, T & data, ...)`

Reply:
same as above

Reply:
```
In file included from /ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/header_symlinks/MaterialProperty.h:19,
                 from /ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/header_symlinks/MaterialData.h:12,
                 from /ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/header_symlinks/BlockRestrictable.h:15,
                 from /ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/header_symlinks/MaterialBase.h:14,
                 from /ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/header_symlinks/Material.h:13,
                 from /ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/header_symlinks/ADPiecewiseLinearInterpolationMaterial.h:12,
                 from /ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/src/materials/ADPiecewiseLinearInterpolationMaterial.C:10,
                 from /ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/unity_src/materials_Unity.C:2:
/ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/header_symlinks/DataIO.h: In instantiation of 'void dataStore(std::ostream&, T&, void*) [with T = const std::vector<std::__cxx11::basic_string<char> >; std::ostream = std::basic_ostream<char>]':
/ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/src/materials/MaterialPropertyStorage.C:382:42:   required from here
/ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/header_symlinks/DataIO.h:165:17: error: static assertion failed: Cannot serialize a class that is not trivially copyable!
Write a custom dataStore() template specialization!


   static_assert(std::is_trivially_copyable<T>::value,
                 ^~~
make: *** [/ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build.mk:145: /ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/unity_src/materials_Unity.x86_64-pc-linux-gnu.opt.lo] Error 1
```

We don't instantiate `store/loadHelper(..., const std::vector<T> ...)`

Reply:
K we don't have to do it now, but I think all `dataStores` should be `const`

Reply:
@roystgnr and I have both tried to go down this path... to frustration and not enough time

Reply:
Yeah, @lindsayad; IMHO it's obvious that all dataStores should be const, and it *should* be an easy fix ... but I tried to fix it and everything broke, and before me Logan tried to fix it and everything broke, and before him @permcody tried to fix it and everything broke.  There were some [dire warnings on Slack](https://moosedevelopers.slack.com/archives/C01054VRUEM/p1675802764882089).

You know, with that track record, it would have been funny to have made you fix it before promoting you to MOOSE lead.  Can someone really be the Once and Future King of MOOSE unless they're the only one who can pull the Sword of Const-Correctness from the Stone of Legacy Code?

Reply:
We can add someone else's name as CODEOWNER to all the framework files and see whether they deem it a promotion 😆 

It would be better to have a concrete documented reason why it fails, then to just vaguely wish other developers good luck

Reply:
...is this resolved?

Reply:
> K we don't have to do it now,

This is what I said. As there was another unresolved conversation and there was ongoing discussion on this thread, I did not immediately mark it as resolved. I think I give very punctual reviews and even submit patches when my review has not been addressed. I don't really want to be hastened further